### DISCUSSION TITLE
Sidestepping the polymorphic effect issue with associate types

### DISCUSSION THREAD

**Discussion by mlutze** [flix/flix#6682] [Show and tell]
## Problem

In the Boolean effect set system, effects may not be polymorphic (although effect sets may be). This means one cannot define an effect like the following:

```scala
// ILLEGAL
eff State[a] {
    def get(): a
    def set(x: a): Unit
}
```

Instead one has to monomorphize their effects themselves:

```scala
eff StringState {
    def get(): String
    def set(x: String): Unit
}

eff BoolState {
    def get(): Bool
    def set(x: Bool): Unit
}
```

and no function can be polymorphic over what kind of state it uses. For example, we cannot have

```scala
// ILLEGAL
def useState(x: a): a \ State[a] = {
    do State.set(x);
    do State.get()
}
```

### Solution

We can partially avoid this problem, thanks to associated effects. The idea is that we introduce a type class that wraps the invalid polymorphic effect, and we maintain an associated effect with the class, which is the "real" monomorphic effect.

```scala
class Stateable[a] {
    type Aef: Eff
    pub def get(): a \ Aef
    pub def set(x: a): Unit \ Aef
}
```

Each instance corresponds to one of the monomorphic effects.

```scala
instance Stateable[String] {
    type Aef = StringState
    pub def get(): String \ StringState = do StringState.get()
    pub def set(x: String): Unit \ StringState = do StringState.set(x)
}

instance Stateable[Bool] {
    type Aef = BoolState
    pub def get(): Bool \ BoolState = do BoolState.get()
    pub def set(x: Bool): Unit \ BoolState = do BoolState.set(x)
}
```

And now we can properly write our `useState` function.

```scala
def useState(x: a): a \ Stateable.Aef[a] = {
    Stateable.set(x);
    Stateable.get()
}
```

But can we have polymorphic handlers? It a naive attempt suggests no, because the handler cases must refer to some monomorphic effect:

```scala
// ILLEGAL
def handleState(x: a): a with Stateable[a] = {
    try {
        useState(x)
    } with State[a] { // State[a] is not monomorphic
        def get() = genericGetImpl()
        def set(x: a) = genericSetImpl(x)
    }
}
```

Instead, we can push handlers into the type class as well.

```scala
class Stateable[a] {
    type Aef: Eff
    pub def get(): a \ Aef
    pub def set(x: a): Unit \ Aef
    pub def handle(get: Unit -> a \ efG, set: a -> Unit \ efS, exp: Unit -> b \ efE): b \ efG + efS + (efE - Aef)
}
```

Where each instance handles the proper effect:

```scala
instance Stateable[String] {
    type Aef = StringState
    pub def get(): String \ StringState = do StringState.get()
    pub def set(x: String): Unit \ StringState = do StringState.set(x)
    pub def handle(get: Unit -> String \ efG, set: String -> Unit \ efS, exp: Unit -> b \ efE): b \ efG + efS + (efE - StringState) = {
        try {
            exp()
        } with StringState {
            def get() = get()
            def set(x) = set(x)
    }
}
```

Our `handleState` function then becomes:

```scala
def handleState(x: a): a with Stateable[a] = {
    Stateable.handle(genericGetImpl, genericSetImpl, () => useState(x))
}
```

## Complications

There are several issues with the idea that would need to be addressed.

### Verbosity

That's still a lot of typing to do. We didn't even solve the first problem I mentioned, the tedium of creating a fresh `XState` effect declaration for every type `X`. But I think we can automatically derive a lot of this. As a simple start, we could consider something like an effect template, which can be instantiated with different types. This is something akin to an ML module.

```scala
// EXTENDED LANGUAGE
eff template State[a] { ... }

eff StringState = State[String]
eff BoolState = State[Bool]
```

I suspect much of the complexity here could be hidden behind syntactic sugar, and generation of instances and effects, similar to derivations in today's Flix.

### Polymorphic types

What happens when we want a `State` effect for a list? We cannot have a general `ListState` because that would be polymorphic. Instead we need a different `State` for each `List`.

```scala
// EXTENDED LANGUAGE
eff StringListState = State[List[String]]
eff BoolListState = State[List[Bool]]
```

That's OK, but we run into issues when we try to create the `Stateable` instances, because  e.g. `List[String]` is a complex type, for which creating an instance is not permitted.

```scala
// ILLEGAL
instance Stateable[List[String]] { ... }
```

So instead we might try making a legal, polymorphic instance. But that doesn't work because we need a monomorphic associated effect.

```scala
instance Stateable[List[a]] {
    type Aef = // ??? can't put anything here
}
```

I expect there is a solution to this issue, but I don't see it yet.

**Comment by JonathanStarup** [comment]
could you add a section where you give the intuition of why its impossible in Boolean effects?
